include/fence.h

```diff
@@ -0,0 +1,151 @@
+#pragma once
+
+#include <memory>
+#include <type_traits>
+
+#include "accessor.h"
+#include "buffer_manager.h"
+#include "host_object.h"
+#include "runtime.h"
+#include "task_manager.h"
+
+namespace celerity {
+class distr_queue;
+}
+
+namespace celerity::detail {
+template <typename DataT, int Dims>
+class buffer_fence_promise;
+}
+
+namespace celerity::experimental {
+
+/**
+ * Owned representation of buffer contents as captured by celerity::experimental::fence.
+ */
+template <typename T, int Dims>
+class buffer_snapshot {
+  public:
+	buffer_snapshot() = default;
+
+	buffer_snapshot(buffer_snapshot&& other) noexcept : m_sr(other.m_sr), m_data(std::move(other.m_data)) { other.m_sr = {}; }
+
+	buffer_snapshot& operator=(buffer_snapshot&& other) noexcept {
+		m_sr = other.m_sr, other.m_sr = {};
+		m_data = std::move(other.m_data);
+	}
+
+	id<Dims> get_offset() const { return m_sr.offset; }
+
+	range<Dims> get_range() const { return m_sr.range; }
+
+	subrange<Dims> get_subrange() const { return m_sr; }
+
+	const T* get_data() const { return m_data.get(); }
+
+	inline const T& operator[](const id<Dims> index) const { return m_data[detail::get_linear_index(m_sr.range, index)]; }
+
+	inline detail::subscript_result_t<Dims, const buffer_snapshot> operator[](const size_t index) const { return detail::subscript<Dims>(*this, index); }
+
+  private:
+	template <typename U, int Dims2>
+	friend class detail::buffer_fence_promise;
+
+	subrange<Dims> m_sr;
```

Comment:
Super picky, but I feel like a member variable is sufficiently important/long lived to justify a longer name (i.e. `m_subrange` in this case).

Reply:
Renamed.